core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -116,12 +118,15 @@
     private final Map<Integer, RemoteTask> tasks = new ConcurrentHashMap<>();
 
     // current stage task tracking
-    @GuardedBy("this")
+    private final ReentrantReadWriteLock allTasksLock = new ReentrantReadWriteLock();
+    @GuardedBy("allTasksLock")
     private final Set<TaskId> allTasks = new HashSet<>();
-    @GuardedBy("this")
-    private final Set<TaskId> finishedTasks = new HashSet<>();
-    @GuardedBy("this")
-    private final Set<TaskId> flushingTasks = new HashSet<>();
+
+    private final AtomicInteger finishedTasksCounter = new AtomicInteger();
+    private final AtomicInteger flushingOrFinishedTasksCounter = new AtomicInteger();
+    private volatile boolean flushingTaskWasObserved;
+    private final Set<TaskId> finishedTasks = ConcurrentHashMap.newKeySet();
```

Comment:
if you are using allTasksLock to guard `finishedTasks` or `flushingOrFinishedTasks` then those not need be `ConcurrentHashMap` (HashSet is enough, like before)

Reply:
I am not using `allTasksLock` to guard `finishedTasks` or `flushingOrFinishedTasks`. I use it to guard `allTasks`